sources/api/netdog/src/net_config/v3.rs

```diff
@@ -0,0 +1,106 @@
+//! The `v3` module contains the third version of the network configuration and implements the
+//! appropriate traits.
+
+use super::devices::NetworkDeviceV1;
+use super::{error, Interfaces, Result, Validate};
+use crate::interface_name::InterfaceName;
+use crate::wicked::{WickedInterface, WickedLinkConfig};
+use indexmap::IndexMap;
+use serde::Deserialize;
+use snafu::ensure;
+
+#[derive(Debug, Deserialize)]
+pub(crate) struct NetConfigV3 {
+    #[serde(flatten)]
+    pub(crate) net_devices: IndexMap<InterfaceName, NetworkDeviceV1>,
+}
+
+impl Interfaces for NetConfigV3 {
+    fn primary_interface(&self) -> Option<String> {
+        self.net_devices
+            .iter()
+            .find(|(_, v)| v.primary() == Some(true))
+            .or_else(|| self.net_devices.first())
+            .map(|(n, _)| n.to_string())
+    }
+
+    fn has_interfaces(&self) -> bool {
+        !self.net_devices.is_empty()
+    }
+
+    fn as_wicked_interfaces(&self) -> Vec<WickedInterface> {
+        let mut wicked_interfaces = Vec::new();
+        for (name, config) in &self.net_devices {
+            let interface = WickedInterface::from((name, config));
+
+            // If config is a Bond, we will generate the interface configuration. For each interface:
+            // WickedInterface::new(name), add that to wicked_interfaces
+            if let NetworkDeviceV1::BondDevice(b) = config {
+                for device in &b.interfaces {
+                    let mut wicked_sub_interface = WickedInterface::new(device.clone());
+                    wicked_sub_interface.link = Some(WickedLinkConfig {
+                        master: name.clone(),
+                    });
+
+                    wicked_interfaces.push(wicked_sub_interface)
+                }
+            }
```

Comment:
This is correct, I should probably word this better. As it stands, we generate the entire xml structure for each interface within a bond because wicked is quite specific about what that looks like and we have all the information to do it on behalf of the user. But this is the downside of doing this generation of xml for the user: what to do with configuration the user has provided us. We can attempt to use what they gave us, throw it away and warn them, try to merge them somehow and hope it's valid, or refuse to proceed and make them resolve the conflicting configuration. I opted for the one that forces the user to resolve this, rather than be surprised by the decision we make.